rfcs/0049-flakes.md

```diff
@@ -0,0 +1,709 @@
+  to rely on the Nix search path (e.g. `import <nixpkgs>`) or to use
+  `fetchGit` or `fetchTarball`. The former has poor reproducibility,
+  while the latter is bad UX because of the need to manually update
+  Git hashes to update dependencies.
+
+* `nix-channel` needs a replacement: channels are hard to create,
+  users cannot easily pin specific versions of channels, channels
+  interact in *ad hoc* ways with the Nix search path, and so on.
+
+The flakes mechanism seeks to address all these problems. It can be
+seen as a "Cargo/NPM/... for Nix" except that it's built into Nix
+rather than a separate tool.
+
+# Detailed design
+[design]: #detailed-design
+
+## Flakes
+
+A flake is a Git repository that contains a file named `flake.nix` in
+the root directory. (In the future, there might be other types of
+flakes, such as Mercurial repositories or tarballs.) `flake.nix`
+specifies some metadata about the flake such as dependencies (called
+*inputs*), as well as its *outputs* (the Nix values such as packages
+or NixOS modules provided by the flake).
+
+As an example, below is the `flake.nix` of
+[`dwarffs`](https://github.com/edolstra/dwarffs) (a FUSE filesystem
+for automatically fetching DWARF debug symbols by ELF build ID). It
+depends on the Nixpkgs flake and provides a package (i.e. an
+installable derivation) and a NixOS module.
+
+```
+{
+  name = "dwarffs";
+
+  edition = 201906;
+
+  description = "A filesystem that fetches DWARF debug info from the Internet on demand";
+
+  inputs = [ "nixpkgs" ];
+
+  outputs = inputs: rec {
+    packages.dwarffs =
+      with inputs.nixpkgs.packages;
+      with inputs.nixpkgs.builders;
+      with inputs.nixpkgs.lib;
+
+      stdenv.mkDerivation {
+        name = "dwarffs-0.1.${substring 0 8 inputs.self.lastModified}";
+
+        buildInputs = [ fuse nix nlohmann_json boost ];
+
+        NIX_CFLAGS_COMPILE = "-I ${nix.dev}/include/nix -include ${nix.dev}/include/nix/config.h -D_FILE_OFFSET_BITS=64";
+
+        src = inputs.self;
+
+        installPhase =
+          ''
+            mkdir -p $out/bin $out/lib/systemd/system
+
+            cp dwarffs $out/bin/
+            ln -s dwarffs $out/bin/mount.fuse.dwarffs
+
+            cp ${./run-dwarffs.mount} $out/lib/systemd/system/run-dwarffs.mount
+            cp ${./run-dwarffs.automount} $out/lib/systemd/system/run-dwarffs.automount
+          '';
+      };
+
+    nixosModules.dwarffs = import ./module.nix inputs;
+
+    defaultPackage = packages.dwarffs;
+
+    checks.build = packages.dwarffs;
+  };
+}
+```
+
+A flake has the following attributes:
+
+* `edition`: A number that specifies the version of the flake
+  syntax/semantics to be used. This allows the interpretation of
+  flakes to change in the future. It also enables some evolution of
+  the Nix language; for example, the Nix files in the flake could be
+  parsed using a syntax determined by the edition. The only currently
+  allowed value is `201906`. Nix rejects flakes with an unsupported
```

Comment:
Why use YYYYMM instead of just a number starting at `0` or `1`?

Reply:
It does have the advantage that you can immediately see the age and that parsing it as integer version is strictly monotonic.

Reply:
I'm not sure the age conveys any meaning (vanity metric) and there doesn't seem to be any requirement for versions to be non-monotonic. KISS?

Reply:
I did think about using a simple counter, but using a date seems nicer to users. For example `edition = 3` doesn't really tell you anything unless you've memorized what each edition is. Whereas I more or less know what's in C++11 or Rust 2019.

It's questionably whether the month is really useful though. If we're sure we're never going to change flakes or the language more than once a year we could drop it :-)

Reply:
We can assume the value of the epoch is a non-monotonic, opaque integer which always increases.

Reply:
Surely it'd be a *monotonic*, strictly increasing, opaque integer?

Reply:
My experience from french as well as my reading of the wiktionary both tell me that “strictly increasing” already implies “monotonic”, and that this word is thus jargon that would be better replaced by just “strictly increasing” for anything related to this, as it's both more precise and easier to understand. (something that always decreases or stays constant is monotonic too)